web-registry/src/lib/tracker/types.ts

```diff
@@ -0,0 +1,26 @@
+export interface LoginEvent {
+  date: string;
+  account: string;
+}
+
+export interface Buy1Event {
+  url: string;
+  cardType?: string;
+  buttonLocation: string;
+  projectName?: string | null;
+  projectId?: string | null;
+  creditClassName?: string;
+  creditClassId?: string | null;
+}
+
+export interface Buy2Event {
+  url: string;
+  price?: string;
+  batchDenom?: string;
+  projectName?: string | null;
+  projectId?: string | null;
+  creditClassId?: string;
+  quantity?: number;
+  currencyDenom?: string;
+  retirementAction?: string;
+}
```

Comment:
One comment about these interfaces for the types of events we are going to track. If we add a static property to each of these event types, such as `eventType`, or to continue with `eventName`, like so:

When we build the data to track, as soon as we set this static property, then the rest of the properties that correspond to the type are **inferred**. This way there is no need to import the types at the place where the track function is used.

```
export interface LoginEvent {
  eventName: 'login';                 <<<< Here, for the magic!
  date: string;
  account: string;
}

export interface Buy1Event {
  eventName: 'buy1';                  <<<< Here, for the magic!
  url: string;
  cardType?: string;
  buttonLocation: string;
  projectName?: string | null;
  projectId?: string | null;
  creditClassName?: string;
  creditClassId?: string | null;
}

export interface Buy2Event {
  eventName: 'buy2';                  <<<< Here, for the magic!
  url: string;
  price?: string;
  batchDenom?: string;
  projectName?: string | null;
  projectId?: string | null;
  creditClassId?: string;
  quantity?: number;
  currencyDenom?: string;
  retirementAction?: string;
}
```

If to this improvement, we add another twist to the `useTracker` hook, we can have the following implementation that is very **powerful** (a single input `EventData` that is the union of `EventName` and `EventPayload`)

```
// lib/tracker/types.ts

export type EventData = LoginEvent | Buy1Event | Buy2Event;
```
```
// lib/tracker/useTracker.ts

import { useCallback } from 'react';
import { useTrack } from 'use-analytics';

import { EventData } from './types';

interface UseTracker {
  track: (eventData: EventData) => void;
}

export const useTracker = (): UseTracker => {
  const track = useTrack();

  // our custom tracker function that we expose...
  const tracker = useCallback(
    (eventData: EventData) => {
      const { eventName, ...restEventData } = eventData;
      track(eventName, restEventData);
    },
    [track],
  );

  return {
    track: tracker;
  };
};

```

Reply:
@wgwz the ideal would be to link `eventName` and `eventPayload` in the same data structure, so by reporting only the `eventName` property, the TS itself would infer the corresponding type and there is no possibility of passing the part of the payload that does not correspond to the type. 

Right now as it is, with separate `eventName` and `eventPayload`, nothing prevents us from passing a payload that doesn't correspond to the event type, something like: `track('login', dataBuy1Event)`

Reply:
i think i might prefer an approach that uses a generic, i.e.: 

```typescript
function tracker<T>(eventName: string, eventData: T): T {
 return eventData;
}

interface LoginEvent {
    account: string;
}

const result = tracker<LoginEvent>('login', {account: "regen12345"});

console.log(result);
```

i don't like stuffing the event name into the event data. i prefer having the event name as the first argument to the function. i think it is a distinct and important argument to this function that having it's own position leads to more readable code.

with a generic we will still get type safety: 

![Screen Shot 2022-11-15 at 10 14 34 AM](https://user-images.githubusercontent.com/10120306/201955452-dee6c575-45e8-433f-8a7e-f51270020722.jpg)

Reply:
@wgwz mm, I understand your proposal and it is an option, but I think that the other proposal could be more powerful because it is based on **discriminated unions**, which is a powerful technique... I think that in this case we don't need to create a generic function, but our custom wrapper, which is very specific since it expects our custom event type data that the function itself should know beforehand (its not a generic modulo, but our custom tracker). Creating event types that combine `eventName` (understood as `type` or `kind` property, more idiomatic) with the specific data of the event type (`eventPayload`) is not stuffing the data structures at all!... this allows discriminated unions (related to algebraic types, set theory, etc), and in the code editor type narrowing when using the function...

* https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions
* https://blog.logrocket.com/understanding-discriminated-union-intersection-types-typescript/#discriminated-union

@blushi @flagrede @haveanicedavid @mhagel any thoughts on this?

Reply:
@wgwz also with that approach with generics i think that nothing keep you safe from doing something like this, mixing buy1 name and loginData:

```
tracker<LoginEvent>('buy1', loginData);
```

Reply:
I like when code is easily searchable. So, if I want to find what line in the code has the 'buy1' event, I can search for it. It sounds like this could be accomplished either way.

Reply:
@victordelval we can protect ourselves from that situation by doing it like this:

```typescript
function tracker<S, T>(eventName: S, eventData: T): T {
 return eventData;
}

interface LoginEvent {
    account: string;
}

const result = tracker<'login', LoginEvent>('login', {});

console.log(result);
```

Reply:
There are some interesting ideas here, but maybe we are at risk of over-engineering this for a simple tracking use case. 

Personally, as a dev, I would expect to be able to add a tracking event wherever I want, with a custom payload each time. I'm not sure I follow everything that was suggested here, but that is the key goal of this task.

Maybe we should get this done with a simple implementation, and remember this conversation for future cases. Or we could make a new issue for type-safety.

@wgwz @victordelval

Reply:
yeah I see pros and cons for both approaches... although I have a preference for @victordelval's approach. Since we do know the events we're expecting, I think it's best practice to have to define a new interface in order to support a new event type. We could use `$type` instead of `eventName`.
Also the "generics" approach makes the code a bit repetitive: `tracker<'login', LoginEvent>('login', { ... });` and it doesn't enforce the association of a particular event type with the event payload, you could still do `tracker<'buy1', LoginEvent>('buy1', { ... });` for instance.

Reply:
> it doesn't enforce the association of a particular event type with the event payload, you could still do tracker<'buy1', LoginEvent>('buy1', { ... }); for instance.

i could do the same thing if `LoginEvent` included `eventName`, it would just require modifications across different files, rather than at the call site. i'd argue it's a version of "security by obscurity". maybe i am missing something though.

> We could use $type instead of eventName

@blushi could you explain what the $type is here?

Reply:
> @blushi could you explain what the $type is here?

just a suggestion for a different key name, similarly to what we have already in the @regen-network/api generated types

Reply:
i think given the feedback, i would like to merge this with the generics approach. we can create a separate issue if refactoring and using the approach that @victordelval outlined becomes favorable in the future. i just finished manually testing the added metadata to these events and it's working as intended. would prefer to wait for more changes and another round of testing, in favor getting this deployed, since it does have some important updates for @erikalogie and the marketing team in analytics tools. i.e. error messages when we have buy failures!

Reply:
sounds good, given this comment: https://github.com/regen-network/regen-web/pull/1551#discussion_r1021449051, that could possibly be addressed as part of regen-network/regen-registry#1400